README.md
# screensaver.embellir

*PLEASE NOTE* this is the very experimental branch which tries to use a
built-in http server instead of writing to disk. This might not work well with
Kodi.

A screensaver for Kodi

Features:

* Time, date, and weather visualizations in unusual, non-textual formats
* Both round and square format visualizations
* Adjustable colors
* More data visualzations planned

Currently this is BETA level software and under active development. Use at your own risk.

### Screenshot Gallery

Square format:

![large square screenshot](https://raw.githubusercontent.com/robbieh/screensaver.embellir/master/resources/images/embellir.square.large.jpg)
![small square screenshot 1](https://raw.githubusercontent.com/robbieh/screensaver.embellir/master/resources/images/embellir.square.small.jpg)
![small square screenshot 2](https://raw.githubusercontent.com/robbieh/screensaver.embellir/master/resources/images/embellir.square.small.2.jpg)

Round format:

![large circle screenshot](https://raw.githubusercontent.com/robbieh/screensaver.embellir/master/resources/images/embellir.circle.fullscreen.jpg)
![small circle screenshot](https://raw.githubusercontent.com/robbieh/screensaver.embellir/master/resources/images/embellir.circle.small.jpg)

### Special Thanks

Thanks to enen92 for [Kaster](https://github.com/enen92/screensaver.kaster) and Philip Schmiegelt for [Unary Clock Screensaver](https://github.com/schmiegelt/xbmc.screensaver.unaryclock). Both projects were very useful for me in helping to figure out how to write this for Kodi.

### Etymology

en- 'into'

bellus- 'handsome,pretty,fine'

embellir- to make beautiful

https://www.etymonline.com/word/embellish




changelog.txt
v0.0.1  20190106
- It's alive!


entrypoint.py
# -*- coding: utf-8 -*-
"""
    screensaver.embellir
    Copyright (C) 2017 nundrum

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
"""

from resources.lib.screensaver import Embellir
import xbmcaddon

PATH = xbmcaddon.Addon().getAddonInfo("path")

if __name__ == '__main__':
    screensaver = Embellir(
        'screensaver-embellir.xml',
        PATH,
        'default',
        '',
    )
    screensaver.doModal()
    del screensaver


resources/lib/viz.py
# -*- coding: utf-8 -*-
"""
    screensaver.embellir
    Copyright (C) 2019 nundrum

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
"""
import xbmc
import os
import io
import sys
import json
import requests
import xbmcgui
import xbmcaddon
import xbmcvfs
from . import kodiutils
from .weatherutils import WeatherInfo
from random import randint, shuffle
from .drawutils import Drundle
from PIL import Image, ImageDraw
import time
import calendar
import controller
import operator
import random
from BaseHTTPServer import BaseHTTPRequestHandler, HTTPServer
import threading 

TEMP = xbmc.translatePath("special://temp")

def last_day_of_month(any_day):
    next_month = any_day.replace(day=28) + datetime.timedelta(days=4)  # this will never fail
    return next_month - datetime.timedelta(days=next_month.day)

class EmbellirVizServer(HTTPServer):
    def __init__(self, viz):
        self.viz = viz
        HTTPServer.__init__(self,('127.0.0.1', 48811), EmbellirHTTPRequstHandler)
        xbmc.log("Created http server", xbmc.LOGNOTICE)

    def start_server(self):
        try:
            thread = threading.Thread(target = self.serve_forever)
            thread.start()
            xbmc.log("started thread", xbmc.LOGNOTICE)
        except KeyboardInterrupt:
            self.socket.close()

class EmbellirHTTPRequstHandler(BaseHTTPRequestHandler):

    def log(self,msg):
        t = time.localtime()
        tstr = str(t.tm_min) + ":" + str (t.tm_sec) + " "
        xbmc.log(tstr + str(msg), level=xbmc.LOGNOTICE)

    def do_HEAD(self):
        self.send_response(200)
        self.send_header('Content-type','image/png')
        self.end_headers()

    def do_GET(self):
        xbmc.log(str(time.gmtime().tm_sec) + " serving: " + self.path, xbmc.LOGNOTICE)
        try:
            path = self.path.lstrip('/').split('/')[0]
            self.log("path: " + path)
            if path == "exit":
                self.log("exiting http")
                self.server.stop_server()
            fn = getattr(self.server.viz, path)
            response = fn()
            self.send_response(200)
            self.send_header('Content-length',len(response))
            self.send_header('Content-type','image/png')
            self.end_headers()
            self.wfile.write(response)
        except Exception as e:
            self.send_response(400)
            self.wfile.write(str(e))
            self.end_headers()
            self.log("Failed to get ")
            self.log(str(e))

class EmbellirVizBase():

    def __init__(self, *args, **kwargs):
        pass

    def exit():
        self.exit()

    def drawTime(self):
        raise NotImplementedError

    def drawWeather(self):
        raise NotImplementedError

    def log(self,msg):
        xbmc.log(str(msg), level=xbmc.LOGNOTICE)

class EmbellirVizRound(EmbellirVizBase):

    twelfth = (360/12)
    sixtieth = (360/60)
    weather = WeatherInfo()

    def __init__(self, *args, **kwargs):
        self.size = kwargs['size']
        self.x = kwargs['x']
        self.y = kwargs['y']
        self.config = kwargs['config']
        #self.log(self.weather.weather)
        self.image = Image.new("RGBA",(self.size,self.size), (0,0,0,0))
        self.draw = ImageDraw.Draw(self.image)

    def drawTime(self):
        xbmc.log("in drawTime", xbmc.LOGNOTICE)
        #self.drawCalibrationRings(draw)
        self.draw.rectangle((0,0,size,size),fill=(0,0,0,0))
        draw=self.draw
        size=self.size
        half=size * 0.5
        x=self.x
        y=self.y
        now = time.localtime()

        monthDegrees = now.tm_mon  * self.twelfth
        dayDegrees = now.tm_mday * (360 / calendar.monthrange(now.tm_year, now.tm_mon)[1])
        hourDegrees = (now.tm_hour % 12) * self.twelfth
        minuteDegrees = now.tm_min * self.sixtieth
        secondDegrees = now.tm_sec * self.sixtieth

        #radius
        rm = size * 0.95
        rd = size * 0.85
        rH = size * 0.60
        rM = size * 0.20
        rS = size * 0.10

        drm = Drundle(color=self.config['secondarycolor'], draw=draw)
        drd = Drundle(color=self.config['tertiarycolor'], draw=draw)
        drH = Drundle(color=self.config['primarycolor'], draw=draw)
        drM = Drundle(color=self.config['secondarycolor'], draw=draw)
        drS = Drundle(color=self.config['tertiarycolor'], draw=draw)


        drm.circleArc(x+half,y+half,rm,0,monthDegrees)
        drd.circleArc(x+half,y+half,rd,0,dayDegrees)
        drH.circleArc(x+half,y+half,rH,0,hourDegrees,width=30)
        drM.circleArc(x+half,y+half,rM,0,minuteDegrees,width=20)
        drS.circleArc(x+half,y+half,rS,minuteDegrees,minuteDegrees + secondDegrees,width=10)
        #drp.circleArc(0,0,rd,start,
        #        start+(dayDegrees*outlook)+1,width=10)
        #drc.circleArc(0,0,rd,dayStart+dayDegrees*(F-1)+1,
        #        dayStart+dayDegrees*(F-1)+dayDegrees-1,width=30)
        with io.BytesIO() as content:
            self.image.save(content, format='PNG')
            return content.getvalue()

    def drawCalibrationRings(self):
        draw=self.draw
        size=self.size
        half=size * 0.5
        calibrate = Drundle(color=(125,125,125,125), draw=draw)
        calibrate.translate(half,half)
        calibrate.circleArc(0,0,size*1.0,0,360,width=5)
        calibrate.circleArc(0,0,size*0.9,0,360,width=5)
        calibrate.circleArc(0,0,size*0.8,0,360,width=5)
        calibrate.circleArc(0,0,size*0.7,0,360,width=5)
        calibrate.circleArc(0,0,size*0.6,0,360,width=5)
        calibrate.circleArc(0,0,size*0.5,0,360,width=10)
        calibrate.circleArc(0,0,size*0.4,0,360,width=5)
        calibrate.circleArc(0,0,size*0.3,0,360,width=5)
        calibrate.circleArc(0,0,size*0.2,0,360,width=5)
        calibrate.circleArc(0,0,size*0.1,0,360,width=5)

    def drawWeather(self):
        draw = self.draw
        now = time.localtime()
        size=self.size
        half=size * 0.5
        x=self.x
        y=self.y
        targetRing = 0.4
        ring = size * targetRing
        rrad = ring * 0.5
        dayDegrees = (360 / calendar.monthrange(now.tm_year, now.tm_mon)[1])
        dayStart = dayDegrees * now.tm_mday
        rd = size * 0.85

        drt = Drundle(color=self.config['primarycolor'], draw=draw) #temp
        drp = Drundle(color=self.config['secondarycolor'], draw=draw) #precip
        forecastSizeMax=drt.pointDistance(drt.calcCircle(0,rrad),
                drt.calcCircle(30,rrad))
        #self.log(forecastSizeMax)
        forecastSize=min(forecastSizeMax,
                (size*(targetRing+0.1) - size*(targetRing-0.1)) * 0.5)

        #hourly forecast circles
        drt.translate(half, half)
        drp.translate(half, half)
        hourlyForecast=self.weather.weather['hourly']
        #self.log(hourlyForecast)
        for F in hourlyForecast:
            temp=hourlyForecast[F]['temperature']
            precip=hourlyForecast[F]['precipitation']
            hour = now.tm_hour % 12
            hourDegree = hour * 30 + (F-1) * 30
            #self.log((F,temp,hour))
            xpos,ypos=drt.calcCircle(hourDegree,rrad)
            drt.translate(xpos,ypos)
            drp.translate(xpos,ypos)
            mid = hourDegree - 360
            tempArc = (360/100) * temp * 0.5
            precipArc = (360/100) * precip * 0.5
            drt.circleChord(0,0,forecastSize,0-tempArc+mid,tempArc+mid)
            drp.circleArc(0,0,forecastSize,0-precipArc+mid,precipArc+mid,width=5)
            drt.detranslate()
            drp.detranslate()

        #daily forecast circles
        drc = Drundle(color=(255,0,0,0), draw=draw) #clear
        drc.translate(half, half)
        drD = Drundle(color=self.config['tertiarycolor'], draw=draw) #day
        drD.translate(half, half)

        dailyForecast=self.weather.weather['daily']
        #self.log(dailyForecast)
        for F in dailyForecast:
            high=dailyForecast[F]['high']
            low=dailyForecast[F]['low']
            outlook=dailyForecast[F]['outlook']
            if high==None or low==None or outlook==None:
                continue
            #self.log((low,high,outlook))
            low = min(low,120)
            low = max(low,0)
            high = min(high,120)
            high = max(high,0)
            high=high/120.0
            low=low/120.0
            outlook=outlook/100.0
            #xpos,ypos=drt.calcCircle(hourDegree,rrad)
            start=dayStart+dayDegrees*(F-1)
            drD.circleArc(0,0,rd,start,
                    start+dayDegrees,width=40)
            drc.circleArc(0,0,rd,start+0.5,
                    start+dayDegrees-0.5,width=30)
            #drc.circleArc(0,0,rd,dayStart+dayDegrees*(F-1)+1,
            #        dayStart+dayDegrees*(F-1)+dayDegrees-1,width=30)
            drt.circleArc(0,0,rd,start+1+(dayDegrees*low),
                    start+1+(dayDegrees*high),width=30)
            drp.circleArc(0,0,rd,start+1,
                    start+(dayDegrees*outlook)+1,width=10)
        drt.detranslate()
        drp.detranslate()

class EmbellirVizSquare(EmbellirVizBase):

    def __init__(self, *args, **kwargs):
        self.size = kwargs['size']
        self.x = kwargs['x']
        self.y = kwargs['y']
        self.config = kwargs['config']
        self.image = Image.new("RGBA",(self.size,self.size), (0,0,0,0))
        self.draw = ImageDraw.Draw(self.image)

        self.paddingSize=self.size / 70
        self.boxSize=(self.size - (self.paddingSize * 3)) / 4

        #assign shorter variable names just to make this more readable
        p=self.paddingSize
        b=self.boxSize
        self.hourPositions=[
                (p*2+b*2, p*0+b*0),
                (p*3+b*3, p*0+b*0),
                (p*3+b*3, p*1+b*1),
                (p*3+b*3, p*2+b*2),
                (p*3+b*3, p*3+b*3),
                (p*2+b*2, p*3+b*3),
                (p*1+b*1, p*3+b*3),
                (p*0+b*0, p*3+b*3),
                (p*0+b*0, p*2+b*2),
                (p*0+b*0, p*1+b*1),
                (p*0+b*0, p*0+b*0),
                (p*1+b*1, p*0+b*0),
                ]
        #self.log(self.hourPositions)
        self.minuteSize=self.boxSize / 5
        self.hourSize=self.minuteSize * 3 - self.paddingSize * 2
        self.hourOffset=self.minuteSize + self.paddingSize
        msz=self.minuteSize
        self.minutePositions=[
                [(msz*0,msz*0), (msz*0,msz*1), (msz*0,msz*2), (msz*0,msz*3), (msz*0,msz*4)],
                [(msz*1,msz*0), (msz*1,msz*4)],
                [(msz*2,msz*0), (msz*2,msz*4)],
                [(msz*3,msz*0), (msz*3,msz*4)],
                [(msz*4,msz*0), (msz*4,msz*1), (msz*4,msz*2), (msz*4,msz*3), (msz*4,msz*4)] ]
        self.secondsGrid = [(x,y) for x in range(8) for y in range(8)]
        shuffle(self.secondsGrid)
        self.secondSize=(self.boxSize*2 + self.paddingSize) / 8
        self.secondsOffset=self.boxSize + self.paddingSize
        #self.log("init round")
        #self.log(self.size)
        #self.log(self.x)
        #self.log(self.y)
        #self.log(self.config)
        #self.log(self.paddingSize)
        #self.log(self.boxSize)

    def drawTime(self):
        self.log(self)
        size=self.size
        self.draw.rectangle((0,0,size,size),fill=(0,0,0,0))
        draw = self.draw
        x=self.x
        y=self.y
        p=self.paddingSize
        now = time.localtime()
        hours = now.tm_hour % 12
        minuteList = range(now.tm_min)
        #self.log(now.tm_sec)

        drm = Drundle(color=self.config['primarycolor'], draw=draw)
        drd = Drundle(color=self.config['primarycolor'], draw=draw)
        drH = Drundle(color=self.config['primarycolor'], draw=draw)
        drM = Drundle(color=self.config['secondarycolor'], draw=draw)
        drS = Drundle(color=self.config['tertiarycolor'], draw=draw)

        for h in range(12):
            if h < hours:
                x,y =self.hourPositions[h]
                drH.square(x+self.hourOffset,y+self.hourOffset, self.hourSize, centered=False)
            minuteCount = len(minuteList[:5])
            minuteList = minuteList[5:]
            for m in range(1,minuteCount+1):
                minutePositionList = self.minutePositions[m-1]
                for position in minutePositionList:
                    hx,hy=self.hourPositions[h]
                    x,y=position
                    drM.square(x+hx,y+hy,self.minuteSize,centered=False)

        for s in range(now.tm_sec):
            ss = self.secondSize
            o = self.secondsOffset
            drS.square(self.secondsGrid[s][0]*ss+o,self.secondsGrid[s][1]*ss+o,
                    ss,centered=False)
        with io.BytesIO() as content:
            self.image.save(content, format='PNG')
            return content.getvalue()

    def drawWeather(self,):
        pass



resources/lib/drawutils.py
# -*- coding: utf-8 -*-
"""
    screensaver.embellir
    Copyright (C) 2019 nundrum

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
"""
import xbmc
import os
import math
import json
import requests
import xbmcgui
import xbmcaddon
import xbmcvfs
from . import kodiutils
from random import randint, shuffle
from .screensaverutils import ScreenSaverUtils
from PIL import Image, ImageDraw
import time
import controller

TEMP = xbmc.translatePath("special://temp")

class Drundle():

    def __init__(self, *args, **kwargs):
        self.ref_x = 0
        self.ref_y = 0
        self.translation_stack = []
        self.color = kwargs['color']
        self.draw = kwargs['draw']

    def set_color(self,color):
        self.color = color

    def stack(self):
        return [self.translation_stack, self.ref_x, self.ref_y]

    def translate(self,x,y):
        self.translation_stack.append((self.ref_x,self.ref_y))
        self.ref_x = self.ref_x + x
        self.ref_y = self.ref_y + y

    def detranslate(self):
        self.ref_x, self.ref_y = self.translation_stack.pop()

    def square(self,x,y,size,centered=True):
        s=self
        if centered:
            half=size*0.5
            self.draw.rectangle([s.ref_x+x-half,s.ref_y+y-half, s.ref_x+x+half,s.ref_y+y+half], fill=self.color)
        else:
            self.draw.rectangle([s.ref_x+x,s.ref_y+y, s.ref_x+x+size,s.ref_y+y+size],fill=self.color)

    #very helpful:
    #https://stackoverflow.com/questions/7070912/creating-an-arc-with-a-given-thickness-using-pils-imagedraw
    def arc(self, bbox, start, end, fill, width=1, segments=100):
        """
        Hack that looks similar to PIL's draw.arc(), but can specify a line width.
        """
        segments=int(segments)
        # radians
        #xbmc.log(str(start), level=xbmc.LOGNOTICE)
        start *= math.pi * 1/180
        end *= math.pi * 1/180

        # angle step
        if segments > 0:
            da = (end - start) / segments
        else:
            return

        # shift end points with half a segment angle
        start -= da / 2
        end -= da / 2

        # ellips radii
        rx = (bbox[2] - bbox[0]) / 2
        ry = (bbox[3] - bbox[1]) / 2

        # box centre
        cx = bbox[0] + rx
        cy = bbox[1] + ry

        # segment length
        l = (rx+ry) * da / 2.0

        for i in range(segments):

          # angle centre
          a = start + (i+0.5) * da

          # x,y centre
          x = cx + math.cos(a) * rx
          y = cy + math.sin(a) * ry

          # derivatives
          dx = -math.sin(a) * rx / (rx+ry)
          dy = math.cos(a) * ry / (rx+ry)

          self.draw.line([(x-dx*l,y-dy*l), (x+dx*l, y+dy*l)], fill=fill, width=width)

    def circleArc(self,x,y,size,start,end,width=10,centered=True):
        start -= 90
        end -= 90
        s=self
        if centered:
            half=size*0.5
            self.arc([s.ref_x+x-half,s.ref_y+y-half, s.ref_x+x+half,s.ref_y+y+half], start, end, self.color, width, segments=abs(start-end))
        else:
            self.arc([s.ref_x+x,s.ref_y+y,
                s.ref_x+x+s,s.ref_y+y+s], start, end, width, fill=self.color)

    def circleChord(self,x,y,size,start,end,width=10,centered=True):
        start -= 90
        end -= 90
        s=self
        if centered:
            half=size*0.5
            self.draw.chord([s.ref_x+x-half,s.ref_y+y-half, s.ref_x+x+half,s.ref_y+y+half], start, end, fill=self.color)
        else:
            self.draw.chord([s.ref_x+x,s.ref_y+y,
                s.ref_x+x+s,s.ref_y+y+s], start, end, fill=self.color)


    def calcCircle(self,degrees,radius):
        """return coordinate on cirle of size 'radius',
        corrected such that degree 0 is up"""
        angle = math.radians(degrees-90)
        return(math.cos(angle) * radius,
                math.sin(angle) * radius)

    def pointDistance(self,p1,p2):
        x1,y1=p1
        x2,y2=p2
        return math.hypot(abs(x1 - x2), abs(y1 - y2))

    def log(self,msg):
                xbmc.log(str(msg), level=xbmc.LOGNOTICE)




resources/lib/kodiutils.py
# -*- coding: utf-8 -*-
"""
    screensaver.embellir
    Copyright (C) 2019 nundrum

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
"""
import xbmc
import xbmcaddon
import xbmcgui
import json as json
import os
import sys

ADDON = xbmcaddon.Addon()
ADDON_ID = ADDON.getAddonInfo("id")


def notification(header, message, time=5000, icon=ADDON.getAddonInfo('icon'), sound=True):
    xbmcgui.Dialog().notification(header, message, icon, time, sound)


def show_settings():
    ADDON.openSettings()


def log(message,level):
    prefix = b"[%s] " % ADDON_ID
    formatter = prefix + b'%(name)s: %(message)s'
    try:
        xbmc.log(formatter, level)
    except UnicodeEncodeError:
        xbmc.log(formatter.encode(
            'utf-8', 'ignore'), level)


def get_setting(setting):
    try:
        return ADDON.getSetting(setting).strip().decode('utf-8')
    except AttributeError:
        return ADDON.getSetting(setting).strip()


def set_setting(setting, value):
    ADDON.setSetting(setting, str(value))


def get_setting_as_bool(setting):
    return get_setting(setting).lower() == "true"


def get_setting_as_float(setting):
    try:
        return float(get_setting(setting))
    except ValueError:
        return 0


def get_setting_as_int(setting):
    try:
        return int(get_setting_as_float(setting))
    except ValueError:
        return 0

def get_string(string_id):
    return ADDON.getLocalizedString(string_id).encode('utf-8', 'ignore')

def kodi_json_request(params):
    data = json.dumps(params)
    request = xbmc.executeJSONRPC(data)

    try:
        response = json.loads(request)
    except UnicodeDecodeError:
        response = json.loads(request.decode('utf-8', 'ignore'))

    try:
        if 'result' in response:
            return response['result']
        return None
    except KeyError:
        log("[%s] %s" %
                    (params['method'], response['error']['message']),level=xbmc.LOGWARNING)
        return None

def get_temp_path():
    kodi_temp = os.path.realpath(xbmc.translatePath("special://temp"))
    #check for tmpfs so we can avoid disk thrashing
    if sys.platform == 'linux2':
        try:
            with open("/proc/mounts", "r") as mounts:
                for mount in mounts:
                    fs, path, remainder=mount.split(' ',2)
                    if fs == 'tmpfs':
                        if os.access(path, os.R_OK) and os.access(path, os.W_OK):
                            return path
        except IOError:
            pass
    return kodi_temp






resources/lib/controller.py
#!/usr/bin/python

import threading
import datetime

import xbmc

class Controller(threading.Thread):

    def __init__(self, log_callback, draw_callback, config, redrawInterval):
      super(Controller, self).__init__()
      self.log_callback = log_callback
      self.draw_callback = draw_callback
      self.config = config
      self.redrawInterval = redrawInterval
      self.waitCondition = threading.Condition()
      self._stop = False

    def run(self):
        self.waitCondition.acquire()
        while not self.shouldStop():
             self.now = datetime.datetime.today()
             self.log_callback("controller loop executing")
             self.draw_callback(self.config)
             #TODO
             #if (self.now.second % self.redrawInterval == 0):
             #   self.draw_callback(False)
             #elif (self.showSeconds):
             #   self.draw_callback(True)

             #self.waitFor =  1000000 - self.now.microsecond
             #self.waitCondition.wait(float(self.waitFor) / 1000000)
             self.waitCondition.wait(2)
        self.waitCondition.release()


    def shouldStop(self):
        if (xbmc.abortRequested):
            return True
        return self._stop

    def stop(self):
        self.waitCondition.acquire()
        self._stop = True
        self.waitCondition.notifyAll()
        self.waitCondition.release()


resources/lib/screensaver.py
# -*- coding: utf-8 -*-
"""
    screensaver.embellir
    Copyright (C) 2019 nundrum

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
"""
import xbmc
import os
import json
import requests
import xbmcgui
import xbmcaddon
import xbmcvfs
from . import kodiutils
from random import randint, shuffle
from .screensaverutils import ScreenSaverUtils
from PIL import Image, ImageDraw
import viz
import time
import controller
import threading

PATH = xbmcaddon.Addon().getAddonInfo("path")
IMAGE_FILE = os.path.join(PATH, "resources", "images", "chromecast.json")

class Embellir(xbmcgui.WindowXMLDialog):

    class ExitMonitor(xbmc.Monitor):

        def __init__(self, exit_callback):
            self.exit_callback = exit_callback

        def onScreensaverDeactivated(self):
            try:
                self.exit_callback()
            except AttributeError:
                xbmc.log(
                    msg="exit_callback method not yet available",
                    level=xbmc.LOGWARNING
                )


    def __init__(self, *args, **kwargs):
        self.exit_monitor = None
        self.images = []
        self.set_property()
        self.utils = ScreenSaverUtils()
        self.tmp = kodiutils.get_temp_path()
        self.firstImage = True

    def log(self,msg):
        t = time.localtime()
        tstr = str(t.tm_min) + ":" + str (t.tm_sec) + " "
        xbmc.log(tstr + str(msg), level=xbmc.LOGNOTICE)

    def drawVisualizations(self, config):
        #if clock...
        if self.firstImage:
            t = str(time.gmtime().tm_sec)
            self.log(t + " first")
            self.logo1.setImage("http://localhost:48811/drawTime/" + str(time.time()), useCache=False)
            self.logo1.setVisible(True)
            self.logo2.setVisible(False)
            self.firstImage=False
            self.log(t + " first ended")
        else:
            t = str(time.gmtime().tm_sec)
            self.log(t + " second")
            self.logo2.setImage("http://localhost:48811/drawTime/" + str(time.time()), useCache=False)
            self.logo2.setVisible(True)
            self.logo1.setVisible(False)
            self.firstImage=True
            self.log(t + " second ended")
        #xbmc.sleep(200)

    def onInit(self):
        self._isactive = True
        # Register screensaver deactivate callback function
        self.exit_monitor = self.ExitMonitor(self.exit)
        # Init controls
        self.backgroud = self.getControl(32500)
        self.metadata_line2 = self.getControl(32503)
        self.metadata_line3 = self.getControl(32504)

        self.logo1 = self.getControl(32505)
        self.logo2 = self.getControl(32506)

        config={'size': 900}
        config.update({
            'primarycolor': (
                kodiutils.get_setting_as_int("primary-red"),
                kodiutils.get_setting_as_int("primary-green"),
                kodiutils.get_setting_as_int("primary-blue"),
                kodiutils.get_setting_as_int("primary-alpha")),
            'secondarycolor': (
                kodiutils.get_setting_as_int("secondary-red"),
                kodiutils.get_setting_as_int("secondary-green"),
                kodiutils.get_setting_as_int("secondary-blue"),
                kodiutils.get_setting_as_int("secondary-alpha")),
            'tertiarycolor': (
                kodiutils.get_setting_as_int("tertiary-red"),
                kodiutils.get_setting_as_int("tertiary-green"),
                kodiutils.get_setting_as_int("tertiary-blue"),
                kodiutils.get_setting_as_int("tertiary-alpha"))
            })

        # Grab images
        self.get_images()

        # Instantiate desired viz class
        # if setting is square...
        mode=kodiutils.get_setting_as_int("clock-mode")
        if mode == 0:
            self.viz = viz.EmbellirVizSquare(size=900,x=0,y=0,config=config)
        elif mode == 1:
            self.viz = viz.EmbellirVizRound(size=900,x=0,y=0,config=config)
        if kodiutils.get_setting_as_int("clock-position") == 1:
            self.logo1.setWidth(400)
            self.logo1.setHeight(400)
            self.logo1.setPosition(1920-500,1080-500)
            self.logo2.setWidth(400)
            self.logo2.setHeight(400)
            self.logo2.setPosition(1920-500,1080-500)

        #start http server for selected viz
        self.vizhttp = viz.EmbellirVizServer(self.viz)
        self.vizhttp.start_server()


        #start timer controller for clock image redraw
        self.cont = controller.Controller(self.log, self.drawVisualizations, config, 60)
        self.cont.start()

        # Start Image display loop
        if self.images and self.exit_monitor:
            while self._isactive and not self.exit_monitor.abortRequested():
                rand_index = randint(0, len(self.images) - 1)

                # if it is a google image....
                if "private" not in self.images[rand_index]:
                    if requests.head(url=self.images[rand_index]["url"]).status_code != 200:
                        continue

                    # photo metadata
                    if "location" in self.images[rand_index].keys() and "photographer" in self.images[rand_index].keys():
                        self.metadata_line2.setLabel(self.images[rand_index]["location"])
                        self.metadata_line3.setLabel("%s %s" % (kodiutils.get_string(32001),
                                                                self.utils.remove_unknown_author(self.images[rand_index]["photographer"])))
                    elif "location" in self.images[rand_index].keys() and "photographer" not in self.images[rand_index].keys():
                        self.metadata_line2.setLabel(self.images[rand_index]["location"])
                        self.metadata_line3.setLabel("")
                    elif "location" not in self.images[rand_index].keys() and "photographer" in self.images[rand_index].keys():
                        self.metadata_line2.setLabel("%s %s" % (kodiutils.get_string(32001),
                                                                self.utils.remove_unknown_author(self.images[rand_index]["photographer"])))
                        self.metadata_line3.setLabel("")
                    else:
                        self.metadata_line2.setLabel("")
                        self.metadata_line3.setLabel("")
                else:
                    # Logic for user owned photos - custom information
                    if "line1" in self.images[rand_index]:
                        self.metadata_line2.setLabel(self.images[rand_index]["line1"])
                    else:
                        self.metadata_line2.setLabel("")
                    if "line2" in self.images[rand_index]:
                        self.metadata_line3.setLabel(self.images[rand_index]["line2"])
                    else:
                        self.metadata_line2.setLabel("")
                # Insert photo
                self.backgroud.setImage(self.images[rand_index]["url"])
                # Pop image and wait
                del self.images[rand_index]
                # Sleep for a given ammount of time if the window is active abort is not requested
                # note: abort requested is only called after kodi kills the entrypoint and we need to return
                # as soon as possible
                loop_count = (kodiutils.get_setting_as_int("wait-time-before-changing-image") * 1000) / 200
                for _ in range(0, loop_count):
                    if self._isactive and not self.exit_monitor.abortRequested():
                        xbmc.sleep(200)
                # Check if images dict is empty, if so read the file again
                if not self.images:
                    self.get_images()

    def get_images(self, override=False):
        # Read google images from json file
        self.images = []
        if kodiutils.get_setting_as_int("screensaver-mode") == 0 or kodiutils.get_setting_as_int("screensaver-mode") == 2 or override:
            with open(IMAGE_FILE, "r") as f:
                images = f.read()
            self.images = json.loads(images)
        # Check if we have images to append
        if kodiutils.get_setting_as_int("screensaver-mode") == 1 or kodiutils.get_setting_as_int("screensaver-mode") == 2 and not override:
            if kodiutils.get_setting("my-pictures-folder") and xbmcvfs.exists(xbmc.translatePath(kodiutils.get_setting("my-pictures-folder"))):
                for image in self.utils.get_own_pictures(kodiutils.get_setting("my-pictures-folder")):
                    self.images.append(image)
            else:
                return self.get_images(override=True)
        shuffle(self.images)
        return

    def set_property(self):
        # Kodi does not yet allow scripts to ship font definitions
        skin = xbmc.getSkinDir()
        if "estuary" in skin:
            self.setProperty("clockfont", "fontclock")
        elif "zephyr" in skin:
            self.setProperty("clockfont", "fontzephyr")
        elif "eminence" in skin:
            self.setProperty("clockfont", "fonteminence")
        else:
            self.setProperty("clockfont", "fontmainmenu")
        # Set skin properties as settings
        for setting in ["hide-clock-info", "hide-kodi-logo", "hide-weather-info", "hide-pic-info", "hide-overlay", "show-blackbackground"]:
            self.setProperty(setting, kodiutils.get_setting(setting))
        # Set animations
        if kodiutils.get_setting_as_int("animation") == 1:
            self.setProperty("animation","panzoom")
        return


    def exit(self):
        self.log("stopping")
        self._isactive = False
        # Delete the monitor from memory so we can gracefully remove
        # the screensaver window from memory too
        try:
            exit_response = requests.get('http://localhost:84411/exit')
        except Exception:
            pass
        if self.exit_monitor:
            del self.exit_monitor
        # Finally call close so doModal returns
        self.close()
